Experiment: Replace style hooks coming from `@wordpress/block-editor`

mikejolley opened the pull request:
This PR Fixes #8071 by pulling in style functions from the `@wordpress/block-editor` package for use on our frontend. Style rules come from block `attributes`—we just need to translate them to CSS.

While this does duplicate methods from Gutenberg, it does:

- Remove the `@wordpress/block-editor` dependency which is unsuitable for frontend use and large
- Fix issues like https://github.com/woocommerce/woocommerce-blocks/issues/8910 due to this dependency
- Simplify implementation using a single `useStyleProps` hook

The downside is maintaining the utils inherited from Gutenberg.

While working on this and cleaning up style usage, I came across 3 blocks which had defined frontend scripts that were not used/needed. These need testing also (in instructions below). 

Fixes #8071

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Smoke test product blocks that have global styles applied.
2. Specifically test adding a Stock and a Attribute filter to the template, as well as modifying text color to ensure it still applies to the frontend.
3. Do the same for the price filter block.
4. Ensure at least one product block had custom spacing, color, and typography applied, and that change was both visible in the editor, and on the frontend after saving the template/post.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

This will have a positive impact on the size of dependencies on the frontend.

### Changelog

> Removed the block-editor dependency on the frontend.

mikejolley committed:
Replace all style hooks with useStyleProps hook

mikejolley committed:
Remove border/color/spacing hooks

mikejolley committed:
Style Props Hook

mikejolley commented:
Looks like it didn't remove all dependencies in the report, but I don't see them in my local build. Will investigate, should this be the approach we want to take.

gigitux commented on assets/js/base/utils/get-inline-styles.ts:
```diff
@@ -0,0 +1,163 @@
+ */
+import classnames from 'classnames';
+import { kebabCase } from 'lodash';
+import { getCSSRules } from '@wordpress/style-engine';
```
Should we add `@wordpress/style-engine` in the package.json?

gigitux reviewed (commented):
The code LGTM and it works!

> The downside is maintaining the utils inherited from Gutenberg.

This is my concern too. If some logic will change in the future on Gutenberg/WordPress side, we have to update the logic too :/

nerrad commented on assets/js/base/utils/get-inline-styles.ts:
```diff
@@ -0,0 +1,163 @@
+ * External dependencies
+ */
+import classnames from 'classnames';
+import { kebabCase } from 'lodash';
```
Probably should avoid the use of lodash here given you've worked on extracting it everywhere?

mikejolley replied:
Yeah I was waiting for the other to merge so I can use the new package.

nerrad commented on assets/js/blocks/active-filters/block-wrapper.tsx:
```diff
@@ -11,12 +12,15 @@ import Block from './block';
 
 const BlockWrapper = ( props: Record< string, unknown > ) => {
-	const colorProps = useColorProps( props );
+	const styleProps = useStyleProps( props );
```
Shouldn't this be `props.attributes` (there may be more occurrences of this)?

mikejolley replied:
It does seem so yes, good catch.

mikejolley replied:
Actually I'm not so sure. It seems like `props` are the attributes. I see several of them use:

```
<Block isEditor={ false } attributes={ parseAttributes( props ) } />
```

`parseAttributes` here uses props directory, not `props.attributes`...

mikejolley replied:
~I think I see the issue; its confusing because some blocks have wrappers. I can fix this.~

I was correct initially. The blocks in this context have attributes at the top level, so using `props` here is correct. I'll update the naming to make this more obvious.

mikejolley replied:
I found some blocks (e.g. attribute filters) which don't need frontend.ts and don't need style props to function. I'm removing those because the props passed were in the wrong format.

opr replied:
@mikejolley are you still planning to update the name here? There are a few other instances where `props` is used too which may be worth renaming

mikejolley replied:
@opr No, I don't think they need renaming. `props` contain more than just attributes, and `props` is a generic term anyway.

mikejolley committed:
Make use of `change-case` package

mikejolley commented:
@nerrad I've hesitating pushing too hard because this is work that affects mostly FSE rather than Rubik, but it would be nice to resolve. I assume next steps are to get another review from a FSE team, and raise the issue upstream to make those hooks available. Is that correct?

nerrad commented:
Yes.

mikejolley committed:
Tidy up block wrappers

mikejolley committed:
Attribute filter does not use frontend.ts nor styles within block

mikejolley committed:
Remove frontend from filter blocks and unused styleprops usage

mikejolley committed:
Tidy up variable names so its clearer attributes are not required specifically from blocks

mikejolley commented:
Logged here: https://github.com/WordPress/gutenberg/issues/50334

opr reviewed (approved):
I tested this and it seems to be working fine, I played with a bunch of styling options on the products block and the filter blocks, it seems global styles are working fine for all of them.

I did not do a deep code review as I am very unfamiliar with this side of the project, but I looked at all files and didn't find anything that seems obviously problematic. Maybe one of the store editing teams could do a deeper review of the actual code?

tjcafferkey commented on assets/js/atomic/blocks/product-elements/price/block.tsx:
```diff
@@ -97,6 +86,7 @@ export const Block = ( props: Props ): JSX.Element | null => {
 		<ProductPrice
 			align={ textAlign }
 			className={ wrapperClassName }
+			style={ styleProps.style }
```
Can't this just be `style={ styleProps.style }`?

tjcafferkey replied:
```suggestion
			style={ style }
```

Aljullu commented on assets/js/atomic/blocks/product-elements/title/block.tsx:
```diff
@@ -87,9 +79,7 @@ export const Block = ( props: Props ): JSX.Element => {
+								...styleProps.style,
 						  }
 						: {}
 				}
```
Nit, this can be rewritten as:

```JS
				style={ isFeaturePluginBuild() ? styleProps.style : {} }
```

Aljullu commented on assets/js/atomic/blocks/product-elements/title/block.tsx:
```diff
@@ -113,9 +103,7 @@ export const Block = ( props: Props ): JSX.Element => {
+							...styleProps.style,
 					  }
 					: {}
 			}
```
Same here.

Aljullu commented on assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-title-label-block/block.tsx:
```diff
@@ -19,22 +15,12 @@ type Props = {
-				...spacingProps.style,
-			} }
+			className={ classnames( props.className, styleProps.className ) }
+			style={ { ...styleProps.style } }
```
This could be simplified as well, if I'm not wrong:
```suggestion
			style={ styleProps.style }
```

Aljullu commented on assets/js/blocks/attribute-filter/block-wrapper.tsx:
```diff
@@ -11,14 +12,18 @@ import Block from './block';
+				isString( props.className ) ? props.className : '',
+				styleProps.className
+			) }
+			style={ { ...styleProps.style } }
```
Same:
```suggestion
			style={ styleProps.style }
```

Aljullu commented on assets/js/blocks/active-filters/block-wrapper.tsx:
```diff
@@ -11,14 +12,18 @@ import Block from './block';
+				isString( props.className ) ? props.className : '',
+				styleProps.className
+			) }
+			style={ { ...styleProps.style } }
```
:slightly_smiling_face: 

```suggestion
			style={ styleProps.style }
```

Aljullu reviewed (commented):
This is looking great, thanks for working on this, @mikejolley!

I left some inline comments in places where we can avoid spreading objects. It's a minor change but thought I would mention anyway because it keeps code simpler.

Besides that, the only regression I found is when changing the background color of the Product Title inner block in the All Products block:

`trunk` | This branch
--- | ---
![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/fc573e6c-74c2-4890-9652-3286ea5cee22) | ![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/deeb9238-f179-486b-a18f-f92c7a9e078a)

tjcafferkey commented on assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-cart-button-block/block.tsx:
```diff
@@ -33,10 +33,10 @@ const Block = ( {
 				'wc-block-mini-cart__footer-cart'
 			) }
-			style={ { ...colorProps.style } }
+			style={ { ...styleProps.style } }
```
```suggestion
			style={ styleProps.style }
```

tjcafferkey commented on assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-checkout-button-block/block.tsx:
```diff
@@ -33,11 +33,11 @@ const Block = ( {
 			) }
 			variant={ getVariant( className, 'contained' ) }
-			style={ { ...colorProps.style } }
+			style={ { ...styleProps.style } }
```
```suggestion
			style={ styleProps.style }
```

tjcafferkey commented on assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-title-items-counter-block/block.tsx:
```diff
@@ -4,33 +4,21 @@
 			style={ {
-				...colorProps.style,
-				...typographyProps.style,
-				...spacingProps.style,
+				...styleProps.style,
 			} }
```
```suggestion
style={ ...styleProps.style }
```

tjcafferkey replied:
```suggestion
style={ styleProps.style }
```

tjcafferkey commented on assets/js/blocks/price-filter/block-wrapper.tsx:
```diff
@@ -11,14 +12,18 @@ import Block from './block';
+				isString( props.className ) ? props.className : '',
+				styleProps.className
+			) }
+			style={ { ...styleProps.style } }
```
```suggestion
			style={ styleProps.style }
```

tjcafferkey commented on assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-title-items-counter-block/block.tsx:
```diff
@@ -4,33 +4,21 @@
-				typographyProps.className
-			) }
+			className={ classNames( props.className, styleProps.className ) }
 			style={ {
```
```suggestion
			style={ styleProps.style }
```

tjcafferkey commented on assets/js/blocks/rating-filter/block-wrapper.tsx:
```diff
@@ -11,14 +12,18 @@ import Block from './block';
+				isString( props.className ) ? props.className : '',
+				styleProps.className
+			) }
+			style={ { ...styleProps.style } }
```
```suggestion
			style={ styleProps.style }
```

tjcafferkey commented on assets/js/blocks/stock-filter/block-wrapper.tsx:
```diff
@@ -11,14 +12,18 @@ import Block from './block';
+				isString( props.className ) ? props.className : '',
+				styleProps.className
+			) }
+			style={ { ...styleProps.style } }
```
```suggestion
			style={ styleProps.style }
```

tjcafferkey reviewed (commented):
Looks good and tests well @mikejolley ! I wasn't able to find any regressions that haven't been highlighted by others.

Regarding the code, I saw @Aljullu 's comment:

>I left some inline comments in places where we can avoid spreading objects. It's a minor change but thought I would mention anyway because it keeps code simpler.

I also left a few minor inline comments for opportunities to do this, for consistency more than anything else.

mikejolley committed:
Update assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-title-label-block/block.tsx

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

mikejolley committed:
Update assets/js/blocks/attribute-filter/block-wrapper.tsx

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

mikejolley committed:
Update assets/js/blocks/active-filters/block-wrapper.tsx

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

mikejolley committed:
Update assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-checkout-button-block/block.tsx

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>

mikejolley committed:
Update assets/js/blocks/rating-filter/block-wrapper.tsx

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>

mikejolley committed:
Update assets/js/blocks/stock-filter/block-wrapper.tsx

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>

mikejolley committed:
Update assets/js/blocks/price-filter/block-wrapper.tsx

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>

mikejolley committed:
Update assets/js/blocks/mini-cart/mini-cart-contents/inner-blocks/mini-cart-cart-button-block/block.tsx

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>

mikejolley committed:
Simplify styleprop

mikejolley committed:
Styleprops simplify

mikejolley committed:
Fix withFeaturedItem styles

mikejolley committed:
Like the original hook, flatten props and combine with parsed styles

mikejolley commented:
Thanks for the reviews @opr @Aljullu @tjcafferkey — I've adjusted those spreads, fixed the missing styles in atomic blocks, and retested. Tests are green so merging!
